File: LeetCode/interleaving_string.rb
#  Given s1, s2, s3, find whether s3 is formed by the interleaving of s1 and s2.
#
#  For example,
#  Given:
#  s1 = "aabcc",
#  s2 = "dbbca",
#
#  When s3 = "aadbbcbcac", return true.
#  When s3 = "aadbbbaccc", return false. 
#

class String
    def inter(first, second)
        inter_rec(self, first, second)
    end

    def inter_rec(s, first, second)
        return true if s.empty? && first.empty? && second.empty?
        return false if s.empty?
        return s[0] == first[0] && inter_rec(s[1..-1], first[1..-1], second) ||
               s[0] == second[0] && inter_rec(s[1..-1], first, second[1..-1])
    end

    def inter_dp(first, second)
        match = Array.new(first.size + 1) { |x|
            Array.new(second.size + 1) { |y|
                false
            }
        }
        0.upto(first.size).each { |i|
            0.upto(second.size).each { |j|
                if i == 0 && j == 0
                    match[i][j] = true 
                    next
                end 
                if i == 0 && second[j - 1] == self[j - 1]
                    match[i][j] = match[i][j - 1]
                    next
                end
                if j == 0 && first[i - 1] == self[j - 1]
                    match[i][j] = match[i - 1][j]
                    next
                end
                if first[i - 1] != self[i + j - 1] && second[i - 1] != self[i + j - 1]
                    match[i][j] = match[i - 1][j] || match[i][j - 1]
                    next
                end
                if first[i - 1] == self[i + j - 1] && second[i - 1] != self[i + j - 1]
                    match[i][j] = match[i - 1][j]
                    next
                end
                if first[i - 1] != self[i + j - 1] && second[i - 1] == self[i + j - 1]
                    match[i][j] = match[i][j - 1]
                    next
                end
            }
        }

        match[first.size][second.size]
    end

    private :inter_rec
end

puts "aadbbcbcac".inter("aabcc", "dbbca")
puts "aadbbbaccc".inter("aabcc", "dbbca")

puts "aadbbcbcac".inter_dp("aabcc", "dbbca")
puts "aadbbbaccc".inter_dp("aabcc", "dbbca")

File: GeeksForGeeks/interleaving_string.cs
//  http://www.geeksforgeeks.org/check-whether-a-given-string-is-an-interleaving-of-two-other-given-strings-set-2/
//
//  Given three strings A, B and C.
//  Write a function that checks whether C is an interleaving of A and B.
//  C is said to be interleaving A and B,
//  if it contains all characters of A and B and order of all characters in individual strings is preserved. 

using System;

static class Program
{
    static bool Inter(this String s, String a, String b)
    {
        if (String.IsNullOrEmpty(s))
        {
            return String.IsNullOrEmpty(a) && String.IsNullOrEmpty(b);
        }

        return !String.IsNullOrEmpty(a) && s[0] == a[0] && s.Substring(1).Inter(a.Substring(1), b) ||
               !String.IsNullOrEmpty(b) && s[0] == b[0] && s.Substring(1).Inter(a, b.Substring(1));
    }

    static bool InterDp(this String s, String a, String b)
    {
        var n = a.Length;
        var m = b.Length;
        if (s.Length != n + m)
        {
            return false;
        }

        var interleaved = new bool[n + 1, m + 1];
        interleaved[0, 0] = true;
        for (var i = 1; i <= n && (interleaved[i, 0] = s[i - 1] == a[i - 1] && interleaved[i - 1, 0]); i++) {}
        for (var i = 1; i <= n && (interleaved[0, 0] = s[i - 1] == b[i - 1] && interleaved[0, i - 1]); i++) {}
        for (var i = 1; i <= n; i++)
        {
            for (var j = 1; j <= m; j++)
            {
                interleaved[i, j] = interleaved[i - 1, j] && a[i - 1] == s[i + j - 1] ||
                                    interleaved[i, j - 1] && b[j - 1] == s[i + j - 1];
            }
        }

        return interleaved[n, m];
    }

    static void Main()
    {
        foreach (var x in new [] {
                    Tuple.Create("aadbbcbcac","aabcc","dbbca", true),
                    Tuple.Create("abc","","abc", true),
                    Tuple.Create("aadbbbaccc","aabcc","dbbca", false),
                    Tuple.Create("abc","","ab", false)
                })
        {
            Console.WriteLine("{0} {1} {2}", x.Item1, x.Item1.Inter(x.Item2, x.Item3), x.Item4);
            Console.WriteLine("{0} {1} {2}", x.Item1, x.Item1.InterDp(x.Item2, x.Item3), x.Item4);
        }
    }
}
